### ISSUE TITLE
Figure out Windows issues with installing in parallel

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubygems/rubygems#3719]
This is a follow up ticket to #3448.

That's closed since we should no longer have the flaky test failures in our test suite, but we should still try to figure out the underlying issue.

The issue is that when `bundle install`'ing Gemfile's with git gems under Windows, we sometimes get stuff like this:

```
$  D:/rubyinstaller-2.6.5-1-x64/bin/ruby.exe -Id:/a/rubygems/rubygems/bundler/lib;d:/a/rubygems/rubygems/bundler/spec -rsupport/hax -rsupport/artifice/fail d:/a/rubygems/rubygems/bundler/exe/bundle install --retry 0
Fetching d:/a/rubygems/rubygems/bundler/tmp/1/libs/foo-1.0
Fetching source index from file:///d:/a/rubygems/rubygems/bundler/tmp/1/gems/remote1/
Resolving dependencies...
Using rake 13.0.1
Using foo 1.0 from d:/a/rubygems/rubygems/bundler/tmp/1/libs/foo-1.0 (at branch2@Using b)
Bundle complete! 1 Gemfile dependency, 3 gems now installed.
```

Not the `Using foo 1.0 from d:/a/rubygems/rubygems/bundler/tmp/1/libs/foo-1.0 (at branch2@Using b)` part.

This result in the git reference not being properly checked out, and the bundle not being usable and leading to errors like:

```
$ D:/rubyinstaller-2.6.5-1-x64/bin/ruby.exe -Id:/a/rubygems/rubygems/bundler/lib -w -e require\ \'d:/a/rubygems/rubygems/bundler/lib/bundler\'\ \;\ Bundler.ui.silence\ \{\ Bundler.setup\(\)\ \}'
       '\ \ \ \ \ \ \ \ require\ \'foo\''
       '\ \ \ \ \ \ \ \ puts\ FOO'
       '
d:/a/rubygems/rubygems/bundler/lib/bundler/source/git/git_proxy.rb:241:in `allowed_in_path': The git source d:/a/rubygems/rubygems/bundler/tmp/1/libs/foo-1.0 is not yet checked out. Please run `bundle install` before trying to start your application (Bundler::GitError)
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git/git_proxy.rb:198:in `find_local_revision'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git/git_proxy.rb:68:in `revision'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git.rb:226:in `revision'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git.rb:93:in `install_path'
 	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/path.rb:131:in `expanded_path'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/path.rb:172:in `load_spec_files'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git.rb:201:in `load_spec_files'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/path.rb:105:in `local_specs'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/source/git.rb:168:in `specs'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:768:in `block in converge_locked_specs'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:754:in `each'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:754:in `converge_locked_specs'
 	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:252:in `resolve'
 	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:174:in `specs'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:242:in `specs_for'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/definition.rb:230:in `requested_specs'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/runtime.rb:101:in `block in definition_method'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/runtime.rb:20:in `setup'
	from d:/a/rubygems/rubygems/bundler/lib/bundler.rb:149:in `setup'
 	from -e:1:in `block in <main>'
	from d:/a/rubygems/rubygems/bundler/lib/bundler/ui/shell.rb:136:in `with_level'
 	from d:/a/rubygems/rubygems/bundler/lib/bundler/ui/shell.rb:88:in `silence'
 	from -e:1:in `<main>'
```

Any progress in this can be logged here.

**Comment by hsbt** [comment]
@deivid-rodriguez Do you interested in supporting native Windows platform like mswin or mingw? I think the users of Windows will use WSL2 environment in 2021.

So We should remove to support them now, I think.

**Comment by deivid-rodriguez** [comment]
Will ruby-core stop supporting native Windows? I hope WSL2 will become widely used in the future, but I don't think we should just drop support for native Windows now.

**Comment by hsbt** [comment]
> Will ruby-core stop supporting native Windows?

No, But We didn't run `test-bundler` on mswin and mingw platforms. 

FYI: We skip a lot of tests/specs with mswin platform like https://github.com/ruby/ruby/commit/cc1ea6e0238a36c08a6255a6ac7a68c54ccafa4a .

I don't have strong opinion about this. I will try to run `test-bundler` with mswin platform.

**Comment by deivid-rodriguez** [comment]
Yeah, we skip quite a few specs too. I think that's fine for now, we don't need to make any extra efforts as long as we don't break what already works.

**Comment by deivid-rodriguez** [comment]
This seems very rare now, but it did reappear after enabling parallel installation by default on Windows :(

https://github.com/rubygems/rubygems/runs/3431902441

I guess I'll reopen this.

I'm going to restart that CI, so the log might disappear, so I'll copy it here (same kind of failure as before):

```
Failures:

  1) [TEST_ENV_NUMBER=2] bundle install with git sources when specifying a revision works when the revision is a non-head ref and it was previously downloaded
     Failure/Error:
               raise <<~ERROR
       
                 Invoking `#{cmd}` failed with output:
                 ----------------------------------------------------------------------
                 #{command_execution.stdboth}
                 ----------------------------------------------------------------------
               ERROR

     RuntimeError:


       Commands:
       $ git config --global init.defaultBranch master
       # $? => 0

       $ git init
       Initialized empty Git repository in D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0/.git/
       # $? => 0

       $ git add *
       # $? => 0

       $ git config user.email lol@wut.com
       # $? => 0

       $ git config user.name lolwut
       # $? => 0

       $ git config commit.gpgsign false
       # $? => 0

       $ git commit -m OMG_INITIAL_COMMIT
       [master (root-commit) 63f125a] OMG_INITIAL_COMMIT
        3 files changed, 21 insertions(+)
        create mode 100644 foo.gemspec
        create mode 100644 lib/foo.rb
        create mode 100644 lib/foo/source.rb
       # $? => 0

       $ git rev-parse HEAD
       63f125a8c19f0963b3b75e6e5cf8cc18b25f09e4
       # $? => 0

       $ git checkout master
       Already on 'master'
       # $? => 0

       $ git rev-parse HEAD
       63f125a8c19f0963b3b75e6e5cf8cc18b25f09e4
       # $? => 0

       $ git add *
       # $? => 0

       $ git commit -m BUMP
       [master 10235ba] BUMP
        1 file changed, 1 insertion(+)
       # $? => 0

       $ C:/hostedtoolcache/windows/Ruby/2.5.9/x64/bin/ruby.exe -ID:/a/rubygems/rubygems/bundler/spec -rD:/a/rubygems/rubygems/bundler/spec/support/artifice/fail.rb -rD:/a/rubygems/rubygems/bundler/spec/support/hax.rb D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/bin/bundle install
       Fetching D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0
       Resolving dependencies...
       Using bundler 2.3.0.dev
       Using foo 1.0 from D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0 (at master@10235ba)
       Bundle complete! 1 Gemfile dependency, 2 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       # $? => 0

       $ git checkout master
       Already on 'master'
       # $? => 0

       $ git rev-parse HEAD
       10235ba39162445c8058047af6f09e6ee7cfb243
       # $? => 0

       $ git add *
       # $? => 0

       $ git commit -m BUMP
       [master 8dcd37c] BUMP
        1 file changed, 1 insertion(+), 2 deletions(-)
       # $? => 0

       $ git update-ref -m "Bundler Spec!" refs/bundler/1 master~1
       # $? => 0

       $ git checkout master
       Already on 'master'
       # $? => 0

       $ git branch --list rando
       # $? => 0

       $ git branch rando
       # $? => 0

       $ git checkout rando
       Switched to branch 'rando'
       # $? => 0

       $ git rev-parse HEAD
       8dcd37cd3d1206fa2cd95a72a43c9fe7d0cb83df
       # $? => 0

       $ git add *
       # $? => 0

       $ git commit -m BUMP
       On branch rando
       nothing to commit, working tree clean
       # $? => 1

       $ C:/hostedtoolcache/windows/Ruby/2.5.9/x64/bin/ruby.exe -ID:/a/rubygems/rubygems/bundler/spec -rD:/a/rubygems/rubygems/bundler/spec/support/artifice/fail.rb -rD:/a/rubygems/rubygems/bundler/spec/support/hax.rb D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/bin/bundle install
       Fetching D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0
       Resolving dependencies...
       Using foo 1.0 from D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0 (at refs/bundler/1@Using b)
       Bundle complete! 1 Gemfile dependency, 2 gems now installed.
       Use `bundle info [gemname]` to see where a bundled gem is installed.
       # $? => 0

       Invoking `C:/hostedtoolcache/windows/Ruby/2.5.9/x64/bin/ruby.exe -rD:/a/rubygems/rubygems/bundler/spec/support/hax.rb -w -e require\ \'bundler\'\ \;\ Bundler.ui.silence\ \{\ Bundler.setup\(\)\ \}\ \;\ \ \ \ \ \ \ \ \ require\ \'foo\''
       '\ \ \ \ \ \ \ \ puts\ \"WIN\"\ if\ defined\?\(FOO\)'
       '` failed with output:
       ----------------------------------------------------------------------
       D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git/git_proxy.rb:221:in `allowed_with_path': The git source D:/a/rubygems/rubygems/bundler/tmp/2/libs/foo-1.0 is not yet checked out. Please run `bundle install` before trying to start your application (Bundler::GitError)
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git/git_proxy.rb:191:in `find_local_revision'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git/git_proxy.rb:63:in `revision'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git.rb:228:in `revision'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git.rb:95:in `install_path'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/path.rb:133:in `expanded_path'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/path.rb:174:in `load_spec_files'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git.rb:203:in `load_spec_files'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/path.rb:107:in `local_specs'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/source/git.rb:170:in `specs'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:716:in `block in converge_locked_specs'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:702:in `each'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:702:in `converge_locked_specs'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:256:in `resolve'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:490:in `materialize'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/definition.rb:239:in `specs_for'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/runtime.rb:18:in `setup'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler.rb:149:in `setup'
       	from -e:1:in `block in <main>'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/ui/shell.rb:136:in `with_level'
       	from D:/a/rubygems/rubygems/bundler/tmp/2/gems/system/gems/bundler-2.3.0.dev/lib/bundler/ui/shell.rb:88:in `silence'
       	from -e:1:in `<main>'
       ----------------------------------------------------------------------
     # ./spec/support/helpers.rb:208:in `sys_exec'
     # ./spec/support/helpers.rb:136:in `ruby'
     # ./spec/support/helpers.rb:64:in `run'
     # ./spec/install/gemfile/git_spec.rb:282:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:100:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:100:in `block (3 levels) in <top (required)>'
     # ./spec/support/helpers.rb:352:in `block in with_gem_path_as'
     # ./spec/support/helpers.rb:366:in `without_env_side_effects'
     # ./spec/support/helpers.rb:348:in `with_gem_path_as'
     # ./spec/spec_helper.rb:99:in `block (2 levels) in <top (required)>'
     # ./spec/support/rubygems_ext.rb:99:in `load'
     # ./spec/support/rubygems_ext.rb:99:in `gem_load_and_activate'
     # ./spec/support/rubygems_ext.rb:18:in `gem_load'

Finished in 51 minutes 54 seconds (files took 0 seconds to load)
2786 examples, 1 failure, 121 pending

Failed examples:

rspec ./spec/install/gemfile/git_spec.rb:254 # [TEST_ENV_NUMBER=2] bundle install with git sources when specifying a revision works when the revision is a non-head ref and it was previously downloaded
```

**Comment by deivid-rodriguez** [comment]
Issue hasn't reappeared for months in CI, so I'll close this again. This change to the Windows default will be released with bundler 2.3.0, so we might need to reopen this once end users start using it, hopefully not!
